### PR TITLE
Dynamic scheduling

### PR DESCRIPTION
**Pull request by nateyazdani**
Merge the feature branch.

### CODE REVIEW

**Comment by rtnpro** [comment]
@yazman 

I tried `fedmsg-hub` on this branch. I am getting an error in `moksha` during running `fedmsg-hub`:

```
[2015-07-10 20:16:03][moksha.hub   ERROR] Failed to init <generator object factory at 0x7fad22e02c80> producer.
Traceback (most recent call last):
  File "build/bdist.linux-x86_64/egg/moksha/hub/hub.py", line 404, in __init_producers
    producer_obj = producer_class(self)
TypeError: 'generator' object is not callable
```

Could you have a look at it?

File: statscache/producer.py [hunk 1]
```diff
@@ -13,81 +14,94 @@
...
+            """ Dynamically generated class for a specific frequency """
+            pass # we need to programmatically set class attributes
+        frequency = statscache.frequency.Frequency(interval)
+        StatsProducerAnon.frequency = frequency
+        StatsProducerAnon.plugin_classes = plugin_classes
+        StatsProducerAnon.__name__ = 'StatsProducer' + str(frequency)
+        yield StatsProducerAnon
+    return
+
+producers = factory()
```
Comment (rtnpro) [inline statscache/producer.py:115]: Is it compatible with `moksha`?

```
Traceback (most recent call last):
  File "build/bdist.linux-x86_64/egg/moksha/hub/hub.py", line 404, in __init_producers
    producer_obj = producer_class(self)
TypeError: 'generator' object is not callable
```

File: statscache/producer.py [hunk 2]
```diff
@@ -13,81 +14,94 @@
...
+        log.info("%s called with %i cached items." % (type(self).__name__, n))
 
-        for plugin in self.plugins.values():
-            log.info("  Calling %r" % plugin.name)
+        for plugin in self.plugins:
+            log.info("  Calling %r" % plugin.ident)
             session = self.make_session()
             try:
-                plugin.handle(session, bucket)
+                plugin.handle(session, timestamp, cache)
```
Comment (rtnpro) [inline statscache/producer.py]: Also, can we get rid of the `timestamp` in the plugin's `handle` call and signatures. The vision was for plugins to be stateless, and use the timestamp from fedmsg message.

Comment (nateyazdani) [inline statscache/producer.py]: Sure, I'll push that change to this branch. You may have seen my [prior response](https://github.com/fedora-infra/statscache_plugins/commit/db2ab8e78cfb55a59c4843b7206406bddff265b1#commitcomment-12106627) about this , where I explained my rationale. Essentially, I thought that plugins were to assume that messages received in `handle()` arrived during the interval ending on `self.frequency.last()`, but now that I understand the statelessness of the interface, I do agree with your point :+1:

File: statscache/producer.py (hunk 1 above)
Comment (nateyazdani) [inline statscache/producer.py:115]: Yes, it is. But, and this is something I should've stated more clearly, it requires [this change](https://github.com/mokshaproject/moksha/pull/29), which is currently only available direct from the git repo. When @ralphbean returns from his vacation, I can work with him to get the version on PyPI updated (he's the package owner).

**Comment by ralphbean** [comment]
:+1: to merge, here.

**Pull request merged**